src/ripple/app/paths/impl/DirectStep.cpp

```diff
@@ -747,39 +758,64 @@ DirectStepI<TDerived>::validFwd (
 // Returns srcQOut, dstQIn
 template <class TDerived>
 std::pair<std::uint32_t, std::uint32_t>
-DirectStepI<TDerived>::qualities (
-    ReadView const& sb,
-    bool srcRedeems,
-    bool fwd) const
+DirectStepI<TDerived>::qualitiesSrcRedeems(
+    ReadView const& sb) const
 {
-    if (srcRedeems)
-    {
-        if (!prevStep_)
-            return {QUALITY_ONE, QUALITY_ONE};
+    if (!prevStep_)
+        return {QUALITY_ONE, QUALITY_ONE};
 
-        auto const prevStepQIn = prevStep_->lineQualityIn (sb);
-        auto srcQOut = static_cast<TDerived const*>(this)->quality (
-            sb, /* src quality out */ false);
+    auto const prevStepQIn = prevStep_->lineQualityIn(sb);
+    auto srcQOut =
+        static_cast<TDerived const*>(this)->quality(sb, QualityDirection::out);
 
-        if (prevStepQIn > srcQOut)
-            srcQOut = prevStepQIn;
-        return {srcQOut, QUALITY_ONE};
-    }
-    else
-    {
+    if (prevStepQIn > srcQOut)
+        srcQOut = prevStepQIn;
+    return {srcQOut, QUALITY_ONE};
+}
+
+// Returns srcQOut, dstQIn
+template <class TDerived>
+std::pair<std::uint32_t, std::uint32_t>
+DirectStepI<TDerived>::qualitiesSrcIssues(
+    ReadView const& sb,
+    DebtDirection prevStepDebtDirection) const
+{
         // Charge a transfer rate when issuing and previous step redeems
-        auto const prevStepRedeems = prevStep_ && prevStep_->redeems (sb, fwd);
-        assert (static_cast<TDerived const*>(this)->verifyPrevStepRedeems (
-            prevStepRedeems));
 
-        std::uint32_t const srcQOut =
-            prevStepRedeems ? transferRate (sb, src_).value : QUALITY_ONE;
-        auto dstQIn = static_cast<TDerived const*>(this)->quality (
-            sb, /* dst quality in */ true);
+        assert(static_cast<TDerived const*>(this)->verifyPrevStepDebtDirection(
+            prevStepDebtDirection));
+
+        std::uint32_t const srcQOut = redeems(prevStepDebtDirection)
+            ? transferRate(sb, src_).value
+            : QUALITY_ONE;
+        auto dstQIn = static_cast<TDerived const*>(this)->quality(
+            sb, QualityDirection::in);
 
         if (isLast_ && dstQIn > QUALITY_ONE)
             dstQIn = QUALITY_ONE;
         return {srcQOut, dstQIn};
+}
+
+// Returns srcQOut, dstQIn
+template <class TDerived>
+std::pair<std::uint32_t, std::uint32_t>
+DirectStepI<TDerived>::qualities(
+    ReadView const& sb,
+    DebtDirection srcDebtDir,
+    StrandDirection strandDir) const
+{
+    if (redeems(srcDebtDir))
+    {
+        return qualitiesSrcRedeems(sb);
+    }
+    else
+    {
+        auto const prevStepDebtDirection = [&] {
+            if (!prevStep_)
```

Comment:
nit: (consistency) in `BookStep` these conditions were flipped (it read: `if (prevStep_) ... `

Reply:
Fixed in: 954c08819